File: Editor/Tools/MoveAssetTool.cs
using UnityEditor;
using McpUnity.Unity;
using McpUnity.Utils;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for moving (or renaming) an asset to a new path in the Unity Asset Database
    /// </summary>
    public class MoveAssetTool : McpToolBase
    {
        public MoveAssetTool()
        {
            Name = "move_asset";
            Description = "Moves or renames an asset from one path to another in the Unity project";
        }

        /// <summary>
        /// Execute the MoveAsset tool with the provided parameters synchronously
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject. Expects 'sourcePath' and 'destinationPath'</param>
        public override JObject Execute(JObject parameters)
        {
            string sourcePath = parameters["sourcePath"]?.ToObject<string>();
            string destinationPath = parameters["destinationPath"]?.ToObject<string>();

            if (string.IsNullOrEmpty(sourcePath) || string.IsNullOrEmpty(destinationPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameters 'sourcePath' and 'destinationPath' must both be provided",
                    "validation_error"
                );
            }

            if (!sourcePath.StartsWith("Assets/") || !destinationPath.StartsWith("Assets/"))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Both 'sourcePath' and 'destinationPath' must start with 'Assets/'",
                    "validation_error"
                );
            }

            if (AssetDatabase.LoadAssetAtPath<UnityEngine.Object>(sourcePath) == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"No asset found at '{sourcePath}'",
                    "not_found_error"
                );
            }

            // AssetDatabase.MoveAsset returns an empty string on success, or an error message on failure
            string error = AssetDatabase.MoveAsset(sourcePath, destinationPath);
            if (!string.IsNullOrEmpty(error))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Failed to move asset from '{sourcePath}' to '{destinationPath}': {error}",
                    "asset_error"
                );
            }

            McpLogger.LogInfo($"[MCP Unity] Moved asset '{sourcePath}' to '{destinationPath}'");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully moved asset to '{destinationPath}'",
                ["sourcePath"] = sourcePath,
                ["destinationPath"] = destinationPath
            };
        }
    }
}

File: Server~/src/tools/moveAssetTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Constants for the tool
const toolName = 'move_asset';
const toolDescription = 'Moves or renames an asset from one path to another in the Unity project';
const paramsSchema = z.object({
  sourcePath: z.string().describe('Assets path of the asset to move (e.g. "Assets/Materials/Red.mat")'),
  destinationPath: z.string().describe('New Assets path for the asset (e.g. "Assets/Materials/Crimson.mat")')
});

/**
 * Creates and registers the Move Asset tool with the MCP server
 * This tool allows moving or renaming an asset in the Unity project
 *
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerMoveAssetTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

/**
 * Handles moving an asset in Unity
 *
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The parameters for the tool
 * @returns A promise that resolves to the tool execution result
 * @throws McpUnityError if the request to Unity fails
 */
async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.sourcePath || !params.destinationPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Both 'sourcePath' and 'destinationPath' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to move asset`
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully moved asset`
    }]
  };
}
